packages/@aws-cdk/aws-cloudfront/lib/response-headers-policy.ts

```diff
@@ -0,0 +1,398 @@
+
+  /**
+   * A list of HTTP methods that CloudFront includes as values for the Access-Control-Allow-Methods HTTP response header.
+   */
+  readonly accessControlAllowMethods: string[];
+
+  /**
+   * A list of origins (domain names) that CloudFront can use as the value for the Access-Control-Allow-Origin HTTP response header.
+   * You can specify `['*']` to allow all origins.
+   */
+  readonly accessControlAllowOrigins: string[];
+
+  /**
+   * A list of HTTP headers that CloudFront includes as values for the Access-Control-Expose-Headers HTTP response header.
+   * You can specify `['*']` to expose all headers.
+   *
+   * @default - no headers exposed
+   */
+  readonly accessControlExposeHeaders?: string[];
+
+  /**
+   * A number that CloudFront uses as the value for the Access-Control-Max-Age HTTP response header.
+   *
+   * @default - no max age
+   */
+  readonly accessControlMaxAge?: Duration;
+
+  /**
+   * A Boolean that determines whether CloudFront overrides HTTP response headers received from the origin with the ones specified in this response headers policy.
+   */
+  readonly originOverride: boolean;
+}
+
+/**
+ * Configuration for a set of HTTP response headers that are sent for requests that match a cache behavior
+ * that’s associated with this response headers policy.
+ */
+export interface ResponseCustomHeadersBehavior {
+  /**
+   * The list of HTTP response headers and their values.
+   */
+  readonly customHeaders: ResponseCustomHeader[];
+}
+
+/**
+ * An HTTP response header name and its value.
+ * CloudFront includes this header in HTTP responses that it sends for requests that match a cache behavior that’s associated with this response headers policy.
+ */
+export interface ResponseCustomHeader {
+  /**
+   * The HTTP response header name.
+   */
+  readonly header: string;
+
+  /**
+   * A Boolean that determines whether CloudFront overrides a response header with the same name
+   * received from the origin with the header specified here.
+   */
+  readonly override: boolean;
+
+  /**
+   * The value for the HTTP response header.
+   */
+  readonly value: string;
+}
+
+/**
+ * Configuration for a set of security-related HTTP response headers.
+ * CloudFront adds these headers to HTTP responses that it sends for requests that match a cache behavior
+ * associated with this response headers policy.
+ */
+export interface ResponseSecurityHeadersBehavior {
+  /**
+   * The policy directives and their values that CloudFront includes as values for the Content-Security-Policy HTTP response header.
+   *
+   * @default - no content security policy
+   */
+  readonly contentSecurityPolicy?: ResponseHeadersContentSecurityPolicy;
+
+  /**
+   * Determines whether CloudFront includes the X-Content-Type-Options HTTP response header with its value set to nosniff.
+   *
+   * @default - no content type options
+   */
+  readonly contentTypeOptions?: ResponseHeadersContentTypeOptions;
```

Comment:
I wonder if this is an instance where we'd prefer to flatten this structure, or otherwise model it a bit differently.

This is an example usage when modeled as-is:

```ts
{
   ...,
   contentTypeOptions: { override: false },
}
```

I'm not sure that's intuitively clear that means "set X-Content-Type-Options header to `nosniff` unless the origin also includes the header". Although I guess since that's the only valid value for the header, maybe it _does_ intuitively mean that. Let me chew on this for a (business) day or so and see if anything else springs to mind as an alternative model.

Reply:
Yeah that was my understanding as well. The override flag would signify if the behavior should override the header if it is already a part of the response from the origin. Therefore I kept it nested. 

More often than not, the customers might use a managed policy for this. But in cases where they would want to customize the behavior, we should give them the full plethora of controls.

Reply:
After giving it some thought (sorry -- much more than a business day or so!) I think this is fine. There are a few alternatives to how it could be modeled, but many try to anticipate usage and simplify for the "normal" case, and I don't (yet) know what that is here. This is functional and reasonably straightforward, so let's roll with it!